resources/puppetlabs/puppetdb/pql/pql-grammar-experimental.ebnf

```diff
@@ -0,0 +1,139 @@
+   generally have support within PDB in one form or another:
+
+   - order by
+   - limit, offset
+   - include_total *)
+
+(* TODO: the follow items need fundamental changes to PDB first:
+
+   - descending into hashes, like facts & resource parameters conveniently
+   - modifying child data returned (ie. document aware projection) *)
+
+(* Top level query, this is the entry point for queries *)
+<query> = from;
+
+(* From - Setting context and providing filtering and projection *)
+from = entity, [<whitespace>], where, [[<whitespace>], extract];
+
+(* Extract - projection *)
+extract = <lbracket>, extractfields, <rbracket>;
+<extractfields> = fieldlist;
+
+(* Filtering *)
+<where> = <lbrace>, [<whitespace>], [expression], [<whitespace>], <rbrace>;
+
+(* Static list of entity types *)
+<entity> = 'facts' |
+           'reports' |
+           'catalogs' |
+           'resources' |
+           'nodes' |
+           'environments' |
+           'factsets' |
+           'fact_paths' |
+           'fact_contents' |
+           'edges' |
+           'events';
+
+(* Expressions
+
+   This structure preserves precedence rules after parsing, so for reference the number
+   after expr is the order of precedence. One should be careful when modifying
+   this structure, and ensure that precedence rules continue to be kept. *)
+<expression> = expr4;
+expr4   = expr3 { <whitespace>, or, <whitespace>, expr4 };
+expr3   = expr2 { <whitespace>, and, <whitespace>, expr3 };
+expr2   = ( not, [<whitespace>], expr2 ) | expr1;
+<expr1> = ( <lparens>, expression, <rparens> ) | condexpression | subquery;
+
+(* Implicit subqueries *)
+subquery = entity, [<whitespace>], where;
+
+(* Conditional expression *)
+condexpression    = (condexpregexp | condexpnumber | condexpmatch | condexpsubquery | condexpnull);
+<condexpregexp>   = field, [<whitespace>], condregexp, [<whitespace>], valueregexp;
+<condexpnumber>   = field, [<whitespace>], condnumber, [<whitespace>], valuenumber;
+<condexpmatch>    = field, [<whitespace>], condmatch, [<whitespace>], valuematch;
+<condexpsubquery> = (field | groupedfieldlist), [<whitespace>], condsubquery, [<whitespace>], query;
+<condexpnull>     = field, [<whitespace>], condnull, [<whitespace>], boolean;
+
+(* Conditional expression parts *)
+groupedfieldlist = <lparens>, [<whitespace>], fieldlist, [<whitespace>], <rparens>;
+<fieldlist>      = field, [ [<whitespace>], <','>, [<whitespace>], fieldlist ];
+<field>          = #'[a-zA-Z_]+';
+
+<condregexp>    = '~';
+<condnumber>    = '>=' | '<=' | '<' | '>';
+<condmatch>     = '=';
+<condsubquery>  = 'in';
+<condnull>      = 'null?';
+
+<valueregexp>   = regexp;
+<valuenumber>   = integer | real;
+<valuematch>    = string | integer | real | boolean;
+
+(* Boolean operators *)
+<and> = <'and'>;
+<or>  = <'or'>;
+not   = <'!'>;
+
+(* Regexp strings *)
+regexp = <regexpquote>, stringwithoutregexpquote, <regexpquote>
+<stringwithoutregexpquote> = #'(?:[^/\\]|\\.)*';
+<regexpquote> = "/"
```

Comment:
Well, I don't know to be honest. There is no difference technically, just that some languages use this syntax (ie. ~ /foo/), and its familiar to some people. Forgetting any technical cost, thats literally where my head was at. I think though, I wouldn't mind so much to drop this, I'm not beholden.

On a similar angle, one could ask a similar related question about single quotes and double quotes ...  right know we have no variable interpolation for example, but we might.

These are the questions that keep me up at night :-). Anyway, I don't mind removing the // ... I think TBH, its probably more common in query languages to use quotes, maybe I'll take a quick survey for arguments sake.